src/rdkafka_partition.c

```diff
@@ -900,8 +907,23 @@ int rd_kafka_retry_msgq(rd_kafka_msgq_t *destq,
                 rd_kafka_msgq_deq(srcq, rkm, 1);
                 rd_kafka_msgq_enq(&retryable, rkm);
 
-                rkm->rkm_u.producer.ts_backoff = backoff;
                 rkm->rkm_u.producer.retries += incr_retry;
+                if (exponential_backoff == rd_true) {
+                        if (rkm->rkm_u.producer.retries > 0)
+                                backoff =
+                                    (1 << (rkm->rkm_u.producer.retries - 1)) *
+                                    retry_ms;
+                        else
+                                backoff = retry_ms;
+                        backoff =
+                            rd_jitter(100 - RD_KAFKA_RETRY_JITTER_PERCENT,
+                                      100 + RD_KAFKA_RETRY_JITTER_PERCENT) *
+                            backoff * 10;
+                        if (backoff > retry_max_ms)
```

Comment:
This comparison is wrong. The statement above sets the unit of backoff to micros (as you multiply by 1000). But here the comparison is with a variable in millis.
See similar comment above this and make the correction.